apps/ca.c

```diff
@@ -728,24 +728,28 @@ int ca_main(int argc, char **argv)
         }
     }
 
-    if (md == NULL && (md = lookup_conf(conf, section, ENV_DEFAULT_MD)) == NULL)
-        goto end;
-
-    if (strcmp(md, "null") == 0) {
+    def_ret = EVP_PKEY_get_default_digest_nid(pkey, &def_nid);
+    /*
+     * EVP_PKEY_get_default_digest_nid() returns 2 if the digest is
+     * mandatory for this algorithm.
+     */
+    if (def_ret == 2 && def_nid == NID_undef) {
```

Comment:
dislike magic numbers, but okay if this is out of scope for this PR.

Reply:
So do I, but I thought it was moving the scope too far to tackle that in this PR.

Reply:
On an additional note: we say that if it returns 2 the digest is mandatory for this algorithm, but we act in that direction only if the return is 2 and the required digest is `NID_undef`.

What if the required default was something else, e.g. SM3 for SM2 or an engine-provided PKEY method with such requirements? Shouldn't it be honored?

Reply:
@romen, perhaps it should - but again I think that starts moving the scope of this PR too far away from the original problem.

Reply:
@mattcaswell I understand that.

My issue is that I'm just not convinced that, in this case, usability can trample on backwards behavioral compatibility.

Is it really necessary that Ed25519/448 should ignore explicit configuration lines/cmdline args?
I ask because to introduce this we are adding a very specific (only for `(EVP_PKEY_get_default_digest_nid() == 2) && (returned_nid == NID_undef)` behavior in the apps, which does not match exactly the documented behavior of the function (*returns 2 if the digest is mandatory for this algorithm*).

Reply:
I'm struggling to see this as a break in backwards compat. If `EVP_PKEY_get_default_digest_nid()` returns 2 then it means the digest is mandatory. If the NID for that mandatory digest is `NID_undef` then what does that mean under the current documentation? About the best one can say is that that behaviour is "undefined". And in practice it makes no difference, because until 1.1.1 came along that never happens! So why does this break backwards compat?

Reply:
> So why does this break backwards compat?

Because if you run the same commands with 1.1.0 apps (with access to Ed25519 through an engine) and 1.1.1 apps, using the same configuration file for the pki settings, the former fails while the second succeeds.

Reply:
Is that a break in backwards compat? There are lots of things that the apps can now do in 1.1.1 that they couldn't do before. Try pointing a 1.1.0 s_client at a TLSv1.3 only server....it will fail. Do the same in 1.1.1 and it will work!

Reply:
More concerning for backwards compat would be if there are things in 1.1.0 that used to work, but now don't.

Reply:
> @romen Because if you run the same commands with 1.1.0 apps (with access to Ed25519 through an engine) and 1.1.1 apps, using the same configuration file for the pki settings, the former fails while the second succeeds.

The whole point of the PR is to succeed where we previously failed. The failure is spurious, the public key algorithm does not use a digest for signing, so we should succeed regardless of any specified digest. Requiring the user to explicitly specify a null digest is a bad UI and breaks existing configuration files and command-line scripts.

If a particular public key algorithm works exclusively with a particular digest or no digest, then we should always use that (or no) digest, and ignore options that provide a digest to use for those key types that could use one of many.

@mattcaswell I think it is time to go ahead and push this PR, I've not seen any substantive objections that prevent moving forward. If someone wants to fine-tune with a follow-on PR we can evaluate that separately.

Reply:
> If a particular public key algorithm works exclusively with a particular digest or no digest, then we should always use that (or no) digest, and ignore options that provide a digest to use for those key types that could use one of many.

@vdukhovni but we don't do this with this changes, only for the "no digest", that's why I am talking about "special case" and not conforming to the documented behavior: if the function returns 2 but it's not asking for the null digest then we still fail!

@mattcaswell I don't mean to stop the PR, I raised my objections and discussion points, which were evaluated, so I am satisfied anyway!

Reply:
I am aware that this PR does not address the case of a single non-null mandatory digest.  We can address that issue when we come to it.  This PR is strictly an improvement, that addresses the immediate issue.  We can generalize later, but feel free to propose a follow-on PR if you have something ready to go.